community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/FreeIdKeeper.java

```diff
@@ -22,90 +22,122 @@
 import java.io.Closeable;
 import java.io.IOException;
 import java.nio.ByteBuffer;
-import java.util.LinkedList;
+import java.util.ArrayList;
+import java.util.List;
 
 import org.neo4j.io.fs.StoreChannel;
 import org.neo4j.kernel.impl.store.UnderlyingStorageException;
 
+import static java.lang.Math.max;
+import static java.lang.Math.toIntExact;
+import static org.neo4j.kernel.impl.store.id.IdContainer.NO_RESULT;
+
 /**
- * Instances of this class maintain a list of free ids with the potential to overflow to disk if the number
- * of free ids becomes too large.
- * This class has no expectations and makes no assertions as to the ids freed. Such consistency guarantees, for
- * example uniqueness of values, should be imposed from users of this class.
+ * Instances of this class maintain a list of free ids with the potential of overflowing to disk if the number
+ * of free ids becomes too large. This class has no expectations and makes no assertions as to the ids freed.
+ * Such consistency guarantees, for example uniqueness of values, should be imposed from users of this class.
+ * <p>
  * There is no guarantee as to the ordering of the values returned (i.e. FIFO, LIFO or any other temporal strategy),
- * primarily because the aggressiveReuse argument influences exactly that behaviour.
- * The {@link StoreChannel} used for persistence can be used by other writers as well. The expectation of this class
- * is that the span of the file from the position it is at when passed at the constructor and forward is available for
- * reads and exclusive writes. Equivalently, instances of this class will never write in the portion of the channel
- * from the beginning until the position it is at when passed at the constructor.
+ * primarily because the aggressiveMode argument influences exactly that behaviour.
+ * <p>
+ * The {@link #aggressiveMode} parameter controls whether or not IDs which are freed during this lifecycle will
+ * be allowed to be reused during the same lifecycle. The alternative non-aggressive behaviour is that the IDs
+ * will only be reused after a close/open cycle. This would generally correlate with a restart of the database.
  */
 public class FreeIdKeeper implements Closeable
 {
-    public static final long NO_RESULT = -1;
-    public static final int ID_ENTRY_SIZE = Long.BYTES;
-    private final LinkedList<Long> freeIds = new LinkedList<>();
-    private final LinkedList<Long> readFromDisk = new LinkedList<>();
+    private static final int ID_ENTRY_SIZE = Long.BYTES;
+
+    private final List<Long> freeIds = new ArrayList<>();
+    private final List<Long> readFromDisk = new ArrayList<>();
```

Comment:
looking into how those used, maybe ArrayDeque will much better here

Reply:
Better in what way?

Reply:
we are interesting in first/last elements most of the time and ArrayDeque will have handy methods that will do exactly that, in addition to that based on Deque interface readers can already understand what kind of operations we are doing with those list of ids

Reply:
Cool, I will update.